src/components/MessageButton.ts

```diff
@@ -0,0 +1,281 @@
+
+	/**
+	 * @param {MessageButtonData|APIButtonComponent} [data] MessageButton to clone or raw data
+	 */
+	public constructor(data?: MessageButtonData | APIButtonComponent) {
+		if (!data) return;
+
+		this.style = MessageButton.resolveStyle(data.style);
+
+		if ('custom_id' in data && typeof data.custom_id !== 'undefined') {
+			this.setCustomID(data.custom_id);
+		} else if ('customID' in data && typeof data.customID !== 'undefined') {
+			this.setCustomID(data.customID);
+		}
+
+		if (typeof data.label !== 'undefined') {
+			this.setLabel(data.label);
+		}
+
+		if (typeof data.emoji !== 'undefined') {
+			this.setEmoji(data.emoji);
+		}
+
+		if (typeof data.url !== 'undefined') {
+			this.setURL(data.url);
+		}
+
+		this.setDisabled(data.disabled ?? false);
+	}
+
+	/**
+	 * Sets the custom ID of this button
+	 * @param {string} customID A unique string to be sent in the interaction when clicked
+	 * @returns {MessageButton}
+	 */
+	public setCustomID(customID: string): this {
+		ow(customID, ow.string.nonEmpty.maxLength(100));
+		ow(this.style, ow.optional.number.inRange(1, 4).message('Cannot set customID on a button that is Link style.'));
+		this.customID = customID;
+		return this;
+	}
+
+	/**
+	 * Sets the interactive status of the button
+	 * @param {boolean} disabled Whether this button should be disabled
+	 * @returns {MessageButton}
+	 */
+	public setDisabled(disabled: boolean): this {
+		ow(disabled, ow.boolean);
+		this.disabled = disabled;
+		return this;
+	}
+
+	/**
+	 * Set the emoji of this button
+	 * @param {EmojiResolvable} emoji The emoji to be displayed on this button
+	 * @returns {MessageButton}
+	 */
+	public setEmoji(emoji: EmojiResolvable): this {
+		const _emoji = MessageButton.resolveEmoji(emoji);
+		ow(_emoji, ow.object.hasAnyKeys('id', 'name'));
+		this.emoji = _emoji;
+		return this;
+	}
+
+	/**
+	 * Sets the label of this button
+	 * @param {string} label The text to be displayed on this button
+	 * @returns {MessageButton}
+	 */
+	public setLabel(label: string): this {
+		ow(label, ow.string.nonEmpty.maxLength(80));
+		this.label = label;
+		return this;
+	}
+
+	/**
+	 * Sets the style of this button
+	 * @param {MessageButtonStyleResolvable} style The style of this button
+	 * @returns {MessageButton}
+	 */
+	public setStyle(style: MessageButtonStyleResolveable): this {
+		this.style = MessageButton.resolveStyle(style);
+		return this;
+	}
+
+	/**
+	 * Sets the URL of this button.
+	 * <note>MessageButton#style must be LINK when setting a URL</note>
+	 * @param {string} url The URL of this button
+	 * @returns {MessageButton}
+	 */
+	public setURL(url: string) {
+		ow(url, ow.string.nonEmpty);
+		ow(this.style, ow.optional.number.equal(5).message('Cannot set URL on a button that is not Link style.'));
+		this.url = url;
+		return this;
+	}
+
+	/**
+	 * Transforms the button to a plain object.
+	 * @returns {APIButtonComponent} The raw data of this button
+	 */
+	public toJSON() {
+		ow(this.style, ow.number);
```

Comment:
Might as well validate it's not any number, but valid numbers, no? Unless we want to allow... forwards-compatibilty with new types